topo/topo.go

```diff
@@ -61,37 +61,65 @@ type Manager struct {
 	links   map[string]*node.Link
 }
 
+type Option func(m *Manager)
+
+func WithKubeClient(c kubernetes.Interface) Option {
+	return func(m *Manager) {
+		m.kClient = c
+	}
+}
+
+func WithTopoClient(c topologyclientv1.Interface) Option {
+	return func(m *Manager) {
+		m.tClient = c
+	}
+}
+
+func WithClusterConfig(r *rest.Config) Option {
+	return func(m *Manager) {
+		m.rCfg = r
+	}
+}
+
 // New creates a new topology manager based on the provided kubecfg and topology.
-func New(kubecfg string, tpb *topopb.Topology) (*Manager, error) {
+func New(kubecfg string, tpb *topopb.Topology, opts ...Option) (*Manager, error) {
 	log.Infof("Creating manager for: %s", tpb.Name)
-	// use the current context in kubeconfig try in-cluster first if not fallback to kubeconfig
-	log.Infof("Trying in-cluster configuration")
-	rCfg, err := rest.InClusterConfig()
-	if err != nil {
-		log.Infof("Falling back to kubeconfig: %q", kubecfg)
-		rCfg, err = clientcmd.BuildConfigFromFlags("", kubecfg)
+	m := &Manager{
+		tpb:   tpb,
+		nodes: map[string]*node.Node{},
+		links: map[string]*node.Link{},
+	}
+	for _, o := range opts {
+		o(m)
+	}
+	if m.rCfg == nil {
+		// use the current context in kubeconfig try in-cluster first if not fallback to kubeconfig
+		log.Infof("Trying in-cluster configuration")
+		rCfg, err := rest.InClusterConfig()
 		if err != nil {
-			return nil, err
+			log.Infof("Falling back to kubeconfig: %q", kubecfg)
+			rCfg, err = clientcmd.BuildConfigFromFlags("", kubecfg)
+			if err != nil {
+				return nil, err
+			}
 		}
+		m.rCfg = rCfg
 	}
-	// create the clientset
-	kClient, err := kubernetes.NewForConfig(rCfg)
-	if err != nil {
-		return nil, err
+	if m.kClient == nil {
+		kClient, err := kubernetes.NewForConfig(m.rCfg)
```

Comment:
Adding a log for creating default kclient or something similar would be helpful here (similar to the case when rCfg is nil)

Reply:
i don't see that being super useful it is just a client - we don't generally "log" creating grpc stub and such

Reply:
I was thinking it could be useful if the user tries to provide one of the newly added options and the log would show that a default value is being used if the intended option wasn't provided.

But I see your point, the in cluster kubecfg logs provide more insight than simply a new kclient was created. SGTM